### PR TITLE
Enforce minimum supported versions of GCC and Clang

### PR DESCRIPTION
**Pull request by phire**
With the move to C++20, and various people wanting to clean up the codebase to use the fancy new features, many PRs have run into problems older versions of clang running on buildbots. 

It would be very helpful to actually define and enforce a minimum version of Clang/GCC so that it's clear what C++20 features contributors are allowed to use. Additionally... enforcing it in cmake gives users clear feedback about what's wrong before they even start, instead of random compile errors. 

We already have an explicit minimum required version of Visual Studio, so it's not too outrageous. And various implicit minimum requirements on other versions.

### Why not feature detection?

That doesn't really answer the question of "what C++20 features are we allowed to use", contributors would have to remember to add feature detection when using new C++20 features, and we would end up implicitly dropping support for versions of compilers without realising it. 

### What versions?

I've chosen Clang 11 as a minimum because that's the version in the latest stable release of FreeBSD. And we have a FreeBSD builder (which currently all the way back on Clang 6, released March 2018, so no wonder it's been unhappy about new C++20 syntax). It's also the minimum for (partial) support of `consteval`.

And GCC 10 seems to be roughly equivalent. Also the minimum for `consteval`. Both were released in 2020.

### CODE REVIEW

**Comment by phire** [comment]
@endrift are you happy updating the freebsd builder to FreeBSD 13.0 or 13.1?

Or should we do something different?

**Comment by endrift** [comment]
Yeah, lemme do that in a bit. Might be a little tricky but I'll try.

File: Contributing.md [hunk 1]
```diff
@@ -173,7 +173,8 @@ Summary:
 - [Classes and Structs](#cpp-code-classes-and-structs)
 
 ## <a name="cpp-code-general"></a>General
-- The codebase currently uses C++17.
+- The codebase currently uses C++20, though not all compilers support all C++20 features
+  - The minimum supported compilers are currently GCC 10, Clang 11 and MSVC 14.32
```
Comment (Pokechu22) [inline Contributing.md:177]: The other bullet points end in a period, so these probably should too.

**Comment by phire** [comment]
For the GCC version, there is an argument to be made that since we don't have a GCC 10.x builder, that we should skip over it and just require the GCC 11.x that our ubuntu uses. 

I'm not sure how much extra distro compatibility GCC 10.x gains us.

 * Ubuntu's releases skips straight from gcc 9 to gcc 11
 * Debian's latest stable release is on gcc 10 (our Debian builders use clang)
 * Fedora is uses gcc 11 on their oldest still supported release.

**Comment by phire** [comment]
FreeBSD apparently updated to Clang 13 with the 13.1 release and didn't mention it in their release notes. But I think we should stick with clang 11 and GCC 10 as our minimums for now (especially with stable Debian still on GCC 10)

The FreeBSD buildbot has been updated to 13.1

**Comment by OatmealDome** [comment]
The latest steamrt used for games (version 1, 'scout') only has gcc-9, which is annoying. It might be possible to just install a newer gcc if we set up a steamrt build worker in the future.

This isn't something we have to worry about right now, anyway.

**Comment by phire** [comment]
Compiler shouldn't be a huge issue. It's a runtime, not a build-time. We don't want to be restrained to the compiler version it uses anyway.

However.... that might put some strain on our minimum libstdc++ version. That's a way less fixable problem. 
Even if we do nothing else, we might want to set up a builder targeting scout now so that we will get errors if people try to use c++20 library features that it doesn't support.

File: Contributing.md [hunk 2]
```diff
@@ -173,7 +173,8 @@ Summary:
 - [Classes and Structs](#cpp-code-classes-and-structs)
 
 ## <a name="cpp-code-general"></a>General
-- The codebase currently uses C++17.
+- The codebase currently uses C++20, though not all compilers support all C++20 features.
+  - The minimum supported compilers are currently GCC 10, Clang 11, and MSVC 14.32.
```
Comment (shuffle2) [inline Contributing.md:177]: maybe just have a pointer to where the check is enforced in code/cmake/etc? we also currently state required VS version in readme and it's annoying to update / probably shouldn't .

**Comment by Minty-Meeo** [comment]
After the minimum GCC version is increased, I have seen a couple of comments around the codebase that can be addressed by this, such as in ShaderGenCommon.h:
```
  // The second template argument is needed to avoid compile errors from ambiguity with multiple
  // enums with the same number of members in GCC prior to 8.  See https://godbolt.org/z/xcKaW1seW
  // and https://godbolt.org/z/hz7Yqq1P5
```

**Comment by AdmiralCurtiss** [comment]
So what's the status here?

**Comment by Minty-Meeo** [comment]
GCC 11 begins GCC's (partial) support for modules.  Whether Dolphin Emulator should or should not begin using modules yet is open for debate.  I personally want to begin experimenting with modules in the Common namespace in the near future.

**Comment by Zopolis4** [comment]
I think GCC 12 is what I'd need for #11057, not sure about clang.

**Comment by Minty-Meeo** [comment]
Oh dang, you beat me to it!  Well best of luck to you, then.

**Comment by phire** [comment]
Personally, I'm a little uneasy about pushing to gcc 12. Debian stable is only shipping with gcc 10.

File: Contributing.md (hunk 2 above)
Comment (Zopolis4) [reply Contributing.md:177]: It's a bit less visual, and its only three numbers that we have to update. Although, I don't see a ton of reason in listing the required compilers in Contributing.md along with Readme.md.

**Comment by Zopolis4** [comment]
I suppose we can merge this PR as is, then I can raise the minimum required versions in #11057,

**Comment by OatmealDome** [comment]
FYI, Valve has given us access to use a newer steam-runtime ("sniper", based on Debian 11), so Linux builds for Steam are no longer a problem.

File: CMakeLists.txt [hunk 3]
```diff
@@ -34,6 +34,33 @@ if (MSVC)
...
+set (Clang_min_version 11)
+set (MSVC_min_version 14.32)
+set (min_vs_version "2022 17.2.3") # contains MSVC_min_version
+
+message(STATUS "Using ${COMPILER} ${CMAKE_CXX_COMPILER_VERSION}")
+
+if ("-" STREQUAL "${${COMPILER}_min_version}-")
+  # TODO change back to just warning
+  # only fatal for buildbot tests
+  message(FATAL_ERROR "Unknown compiler ${COMPILER}, assuming it is new enough")
```
Comment (AdmiralCurtiss) [inline CMakeLists.txt:56]: You probably want to change this still?

**Comment by AdmiralCurtiss** [comment]
I'll ask again, what's the hold up here?

**Comment by Minty-Meeo** [comment]
I think there is a desire for a Steam Deck buildbot.  Also, Android's buildbot still has an outdated stdlib, lacking \<concepts\> last I checked and also std::bit_cast (though it is not alone in that second one).  I would also appreciate an update on this.

**Comment by OatmealDome** [comment]
The Steam Linux runtime is no longer a blocker for this. We've been given access by Valve to a newer version that is based off Debian 11.

**Comment by JosJuice** [comment]
> Also, Android's buildbot still has an outdated stdlib, lacking <concepts> last I checked and also std::bit_cast (though it is not alone in that second one).

The Android buildbot automatically fetches whatever SDK version is specified in the dolphin git repo. Though I suspect that not even the latest Android SDK has a recent enough stdlib for this...

**Comment by Minty-Meeo** [comment]
It is pretty simple to create a "Future" header for std::bit_cast thanks to how old the compiler built-in is for the three major compilers.  See: #11153.  This has also been how I worked around \<concepts\> being MIA for Android.  IMO, this is the way forward until every supported platform gets up to speed with C++20.

**Comment by Minty-Meeo** [comment]
What is the prerequisite or milestone this PR is waiting on?  Are we using C++20 or not?  I'm sorry if this has already been answered in the IRC; I do not have anything set up to see what has been said while I am away.

**Comment by CasualPokePlayer** [comment]
Under the current Dolphin master (45b55f7ccd19a8f71c7b98d14f00668932a3bf50), it seems GCC 10 or Clang 12 (or Apple Clang 13, as reported by Dentomologist on IRC) is required to build Dolphin. GCC 10 got required since 02a967f78625ce226c0590948d49f29d6e35ebe1 and Clang 12 (/Apple Clang 13) goes required since 154cb4f722a9996a4134463e64c85c6bfa25d8bf.

File: CMakeLists.txt [hunk 4]
```diff
@@ -34,6 +34,33 @@ if (MSVC)
   set(CMAKE_CXX_STANDARD_REQUIRED ON)
 endif()
 
+set(COMPILER ${CMAKE_CXX_COMPILER_ID})
+if (COMPILER STREQUAL "GNU")
+  set(COMPILER "GCC") # perfer printing GCC instead of GNU
+endif()
+if (COMPILER MATCHES "Clang")
+  set(COMPILER "Clang") # Simplify all Clang variants to just Clang
```
Comment (CasualPokePlayer) [inline CMakeLists.txt:42]: This wouldn't be a great idea, Apple Clang (what you get with XCode) can effectively be treated as a different compiler, as it lags behind standard Clang in features and doesn't align its versions with standard Clang.

File: CMakeLists.txt [hunk 5]
```diff
@@ -34,6 +34,33 @@ if (MSVC)
...
+endif()
+
+# Enforce minimium compiler versions that support the c++20 features we use
+set (GCC_min_version 10)
+set (Clang_min_version 11)
```
Comment (CasualPokePlayer) [inline CMakeLists.txt:47]: [As I noted here](https://github.com/dolphin-emu/dolphin/pull/10969#issuecomment-1435946039), Dolphin ends up requiring a newer Clang than this, due to https://github.com/dolphin-emu/dolphin/commit/154cb4f722a9996a4134463e64c85c6bfa25d8bf. See "Class types in non-type template parameters" in https://en.cppreference.com/w/cpp/compiler_support.

File: CMakeLists.txt [hunk 6]
```diff
@@ -34,6 +34,33 @@ if (MSVC)
   set(CMAKE_CXX_STANDARD_REQUIRED ON)
 endif()
 
+set(COMPILER ${CMAKE_CXX_COMPILER_ID})
+if (COMPILER STREQUAL "GNU")
+  set(COMPILER "GCC") # perfer printing GCC instead of GNU
```
Comment (CasualPokePlayer) [inline CMakeLists.txt:39]: Note that there is [this (now outdated) GCC check deeper in this CMakeLists](https://github.com/dolphin-emu/dolphin/blob/68d439a4b3b37bd1a0478f473213e0bdb8331e35/CMakeLists.txt#L282-L285), removing that probably should go with this PR.

File: CMakeLists.txt (hunk 4 above)
Comment (phire) [reply CMakeLists.txt:46]: I thought they had reconverged. But looks like that was just a coincidence, and as of Xcode 14.3 beta, they are diverging again, this time in the opposite direction. 

Fine... easy enough to handle the separately.

File: Contributing.md (hunk 2 above)
Comment (phire) [reply Contributing.md]: Ok, I've removed explicit version references in both readme and contributing.

File: CMakeLists.txt (hunk 5 above)
Comment (phire) [reply CMakeLists.txt]: Ok, I've bumped it to Clang 12 (Apple Clang 13)

**Comment by phire** [comment]
Ok, I think I've addressed all review feedback.

File: CMakeLists.txt [hunk 7]
```diff
@@ -38,6 +38,31 @@ if (MSVC)
...
+endif()
+
+# Enforce minimium compiler versions that support the c++20 features we use
+set (GCC_min_version 10)
+set (Clang_min_version 12)
+set (AppleClang_min_version 13.0.0)
+set (min_xcode_version "13.0") # corrosponding xcode version for AppleClang_min_version
+set (MSVC_min_version 14.32)
```
[the remark this answers is not in the document]
Comment (phire) [reply CMakeLists.txt:51]: Hmm... but the inverse also results in having versioning for different compilers split over different places.

I guess we can move it back when we deprecate the visual studio solutions.

**Comment by OatmealDome** [comment]
What's the status on this?

**Pull request merged**